File: built-in plugins dsp/mb-stereo-balance.cpp
/**
 * MB Stereo Balance
 * Category : effect
 * Type     : stereo
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Frequency-dependent stereo balance correction
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_STEREO_BALANCE_H
#define MB_STEREO_BALANCE_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbStereoBalance : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-stereo-balance";
    static constexpr const char* PLUGIN_NAME    = "MB Stereo Balance";
    static constexpr const char* PLUGIN_TYPE    = "stereo";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float balance = 0f;  // range [-1, 1]
    float lowBalance = 0f;  // range [-1, 1]
    float highBalance = 0f;  // range [-1, 1]
    float crossover = 1000f;  // range [200, 5000]
    };

    MbStereoBalance() = default;
    ~MbStereoBalance() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.balance = std::clamp(params.balance, -1f, 1f);
        params.lowBalance = std::clamp(params.lowBalance, -1f, 1f);
        params.highBalance = std::clamp(params.highBalance, -1f, 1f);
        params.crossover = std::clamp(params.crossover, 200f, 5000f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Stereo Balance
        return input;
    }
};

#endif // MB_STEREO_BALANCE_H

File: built-in plugins dsp/mb-stereo-balance.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbStereoBalancePlugin: PluginDefinition = {
  id: "mb-stereo-balance",
  slug: "mb-stereo-balance",
  name: "MB Stereo Balance",
  category: "effect",
  type: "stereo" as any,
  version: "1.0.0",
  description: "Frequency-dependent stereo balance correction",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "balance",
      name: "Balance",
      type: "float",
      defaultValue: 0,
      minValue: -1,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "lowBalance",
      name: "Low Balance",
      type: "float",
      defaultValue: 0,
      minValue: -1,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "highBalance",
      name: "High Balance",
      type: "float",
      defaultValue: 0,
      minValue: -1,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "crossover",
      name: "Crossover",
      type: "float",
      defaultValue: 1000,
      minValue: 200,
      maxValue: 5000,
      automatable: true,
    },
  ],
  defaultPreset: { balance: 0, lowBalance: 0, highBalance: 0, crossover: 1000 },
};

export default MbStereoBalancePlugin;
